ctapipe/coordinates/disp.py

```diff
@@ -0,0 +1,135 @@
+"""
+Helper functions and components for handling coordinate transformations
+during origin reconstruction using the disp method.
+"""
+import warnings
+
+import astropy.units as u
+import numpy as np
+from astropy.coordinates import AltAz, SkyCoord
+from astropy.table import Table
+
+from ctapipe.coordinates import MissingFrameAttributeWarning, TelescopeFrame
+
+from ..containers import ArrayEventContainer, ReconstructedGeometryContainer
+from ..core import Component
+
+__all__ = ["horizontal_to_telescope", "telescope_to_horizontal", "DispConverter"]
+
+
+def horizontal_to_telescope(
+    alt: u.Quantity, az: u.Quantity, pointing_alt: u.Quantity, pointing_az: u.Quantity
+):
+    """Transform coordinates form horizontal coordinates into TelescopeFrame"""
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", MissingFrameAttributeWarning)
+
+        horizontal_coord = SkyCoord(alt=alt, az=az, frame=AltAz())
+        pointing = SkyCoord(alt=pointing_alt, az=pointing_az, frame=AltAz())
+        tel_frame = TelescopeFrame(telescope_pointing=pointing)
+
+        tel_coord = horizontal_coord.transform_to(tel_frame)
+
+    return tel_coord.fov_lon.to(u.deg), tel_coord.fov_lat.to(u.deg)
+
+
+def telescope_to_horizontal(
+    lon: u.Quantity, lat: u.Quantity, pointing_alt: u.Quantity, pointing_az: u.Quantity
+):
+    """Transform coordinates from TelescopeFrame into horizontal coordinates"""
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", MissingFrameAttributeWarning)
+
+        pointing = SkyCoord(alt=pointing_alt, az=pointing_az, frame=AltAz())
+        tel_coord = TelescopeFrame(
+            fov_lon=lon, fov_lat=lat, telescope_pointing=pointing
+        )
+        horizontal_coord = tel_coord.transform_to(AltAz())
+
+    return horizontal_coord.alt.to(u.deg), horizontal_coord.az.to(u.deg)
+
+
+class DispConverter(Component):
+    """Convert (norm, sign) predictions into (alt, az) predictions"""
+
+    # This is a temporary solution for simulations using a single pointing position
+    pointing_altitude = u.Quantity
+    pointing_azimuth = u.Quantity
+
+    def __call__(
+        self, event: ArrayEventContainer, regressor_cls, classifier_cls
+    ) -> None:
+        """Convert and fill in corresponding container"""
+        prefix = regressor_cls + "_" + classifier_cls
+
+        for tel_id in event.trigger.tels_with_trigger:
+            # Maybe NormContainer + SignContainer makes more sense
+            # As is: DispContainer always contains only sign or norm
+            norm = event.dl2.tel[tel_id].disp[regressor_cls].norm
+            valid_norm = event.dl2.tel[tel_id].disp[regressor_cls].norm_is_valid
+            sign = event.dl2.tel[tel_id].disp[classifier_cls].sign
+            valid_sign = event.dl2.tel[tel_id].disp[classifier_cls].sign_is_valid
+
+            if valid_sign:
+                sign = -1 if sign < 0.5 else 1
```

Comment:
Atm `predict` effectively returns one of the resulting scores of `sklearn.XYZclassifier.predict_proba()` which is between 0 and 1. 
This score could be rescaled to be between -1 and 1 or `predict` could return the value -1 or 1 instead. However, returning the class label would be a different behaviour from the `ParticleIdClassifier`, but for sign the score is not important, so it might make sense anyway.
Returning the sign score is nice for some performance plots though.

Reply:
Ah, that means we should change `predict` to `predict_proba` and have a `predict` method that really calls `predict` I would say.

Reply:
I assumed when writing the `Classifier` component, that there isn't a usecase for `predict` and we'd always want the score.

Reply:
The `Classifier` component in `ctapipe/ml/sklearn` does have two functions `predict` and `predict_score`, but the `predict` function of the `ParticleIdClassifier` component in `ctapipe/ml/apply` calls the `predict_score` function of its `Classifier` model. 
So I think renaming `predict` from `ParticleIdClassifier` and `DispClassifier` to `predict_score` and adding a `predict` function to `DispClassifier` which calls the `predict` function of its `Classifier` model, would be the easiest solution.

Reply:
Ah! One level higher. Then no. The `.predict` method of these high-level things should return what we fill into the containers, just as a table. So it is correct that `ParticleIdClassifier.predict` returns the score (because we fill gammaness into the container) but `DispClassifier.predict` should return arrays with `-1` or `1` because that's what is in the `DispContainer`.

I'd also probably wouldn't create two high-level components, just a single one that has the two models:

```
class DispRegressor:
    def __init__(self, ...):
        self.norm_regressor = Regressor(...
        self.sign_classifier = Classifier(... 
```

That would solve the strange "two algorithms" naming as well. Conceptually, this is one 2d regression, not two separate algorithms.

Reply:
The `DispContainer` is filled with the sign score (between 0 and 1) atm. 

Would it make sense to make `self.norm_regressor` and `self.sign_classifier` `Reconstructor`s to keep the functionality of separate quality queries and feature generators for both models?

Reply:
Separate quality queries? An event having only disp or sign is not really useful

Reply:
In regards to the "two algorithms" naming: I was of the impression, that it would be necessary to include the name of the used ML algorithms in the prefix to be able to distinguish if multiple disp models are applied on the same file. Or can/ should this never be the case, anyway?

Reply:
I don't really think that's necessary to have separate feature generators and separate quality queries does not make sense (you would apply the models to different events...).

Since even if you are unsure about the classification you cannot really do anything with that info, I'd rather store the sign in the container (maybe you can store the score in another field if it is really useful)

Reply:
Ok, I understand. Would it still make sense to have this single `DispRegressor` high-level component be a subclass of `Reconstructor`, since `Reconstructor` is designed to work with a single sklearn model, or would it be better to make it a separate `Component`?

I think storing the score in another field is a good idea, event if it's only for performance plots. If we use a separate function to predict the score (e.g. `predict_score`), we could only do so during cross validation. However, if we want to always store the score and the predicted class label, having both `predict` and `predict_score` seems unnecessary, because we could just decide the predicted class label from the score. And if we always store the score and the predicted class label, `predict` would still return what is filled in the container.

Reply:
I implemented a single `DispReconstructor` component whose `predict` function returns `norm`, `sign` and `sign_score`. I also changed the `DispContainer` to contain these three values.

Regarding the "two algorithms" problem: I changed the prefix (and the `algorithm` for the stereo combiner) to be `"disp"` everywhere. This results in a slightly redundant container structure: `event.dl2.tel[tel_id].disp["disp"] = DispContainer`. However using `event.dl2.tel[tel_id].disp = DispContainer` would require changes in the Datawriter and TableLoader, if I am not mistaken. I don't know if such changes would make sense, but I don't think they would be worth it, in any case.